WG serving proposal (#7823)
kubernetes/community: wg-serving/charter.md

# WG Serving Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

Discuss and enhance the support for AI/ML inference workloads in Kubernetes.

> 

Reviewer: We should add @smarterclayton comments during the first call, on what's a Serving workload here in the charter.

Reviewer 2: Yes please!

Reviewer 3: +1 it overlaps a bit with the batch charter as written.

Author: Added one paragraph about it

### In scope

- Gather requirements for serving workloads (inference primarily, but benefiting
  other non-batch use cases where possible) that have broad community alignment
  from practitioners, distros, and vendors. Provide concrete input to other SIGs
  and WGs around needs for identified requirements. Do it in partnership
  with existing ecosystem projects like kServe, Seldon, Kaito, and
  others to identify, extract, or implement common shared problems (like Kueue
  abstracted deferred scheduling for multiple batch frameworks).
- Specific areas of improvement include:
  - Directly improve key kubernetes workload controllers when used with
    accelerators and the most common inference serving frameworks and model
    servers.
  - Explore new projects that improve orchestration, scaling, and load balancing
    of inference workloads and compose well with other workloads on Kubernetes
  - Being able to run serving workloads safely while giving up
    available slack capacity to batch frameworks

### Out of scope

> - Batch inference, which should be covered by WG Batch.

Reviewer: Training?

Author: this refers to batch jobs that will run a batch of inference queries nighty or some similar scenario. So basically something the has characteristics similar to training. Does it make sense or some better language here will help?

Reviewer 4: I guess it might still be worth adding batch training to the list of "out of scope", for additional clarity.

Reviewer 5: Agreed, let's say "training and batch inference are under WG-Batch".

Continuous training may eventually be relevant, but we don't have to call it out here.

> - Addition of new API kinds that serve a specific models. The focus should be on

Reviewer 5: I don't know what this means, or if I agree.  I think we should be focused on CRDs that make inference work consistently on Kube and enable it.  What's the history on this one?

Reviewer 2: I agree with Clayton, I'd drop this point entirely.

Author: removed

    objects with the ability to scale to multi-host 
- To SIG Scheduling and Autoscaling
  - Faster scaling up and down
  - Ability to preempt workloads
- To SIG Node:
  - Runtime support for Pods preemption
  - Runtime support for devices partitioning

## Roles and Organization Management

This wg adheres to the Roles and Organization Management outlined in [wg-governance]
and opts-in to updates and modifications to [wg-governance].

[wg-governance]: /committee-steering/governance/wg-governance.md

Additionally, the wg commits to:

- maintain a solid communication line between the Kubernetes groups and the wider CNCF community;
- submit a proposal to the KubeCon/CloudNativeCon maintainers track;

## Timelines and Disbanding

As a first mandate, the wg will define a roadmap in the first quarter of operation.
We believe there will be a set of features the Working Group can identify and deliver
that will enable the majority of frameworks operate natively on Kubernetes.

There is no expectations that the Working Group will be converted into SIG long term,
however, there is a chance that a separate project or a sizeable sub-component of SIG Apps can be 

> created as a result of a Working Group.

Reviewer 2: I'd like to see clearer exit criteria, maybe something along the deliverables? Iow. after achieving them and the items identified in the roadmap?

Author: I was generally following https://github.com/kubernetes/community/blob/master/wg-batch/charter.md#timelines-and-disbanding.

We don't yet know exact deliverables and as it's said in text we may end up developing a new sub project or it may be an external plugin. It is something we will need to decide.

What specific points are you looking for in this text?

Reviewer 2: Let's try to sketch at least the general exit criteria, for example try adding something like that to the first paragraph:

```
Achieving the aforementioned deliverables, also mentioned in the `In Scope` section,
will allow us to decide when to disband this WG. 
```

Reviewer 3: +1 to @soltysh 's comment. We've had several WGs that have gone on for years making minimal progress with no real exit criteria >_< Even if its fairly plane like that, its still worth it to have written down.

Reviewer 6: Yes, this has been a repeat problem, workgroups are supposed to be scoped and finite, let's try to outline what the exit criteria are now.

Author: added, makes sense. Thank you for clarifying concerns

# WG Serving Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

Discuss and enhance the support for AI/ML inference workloads in Kubernetes.

> By simplified definition, we will say that serving/inference workload is

Reviewer 5: Serving and inference aren't interchangeable - thinking if there is a wording that could work better (this is not suggested wording, just trying to lay out the framing):

> Scope: Discuss and enhance serving workloads on Kubernetes, specifically focused on hardware-accelerated AI/ML inference

> Because serving and batch are both broad categories (pods with restartPolicy=Always vs those with restartPolicy=Never|OnFailure), we focus on the novel challenges of compute-intensive online inference, leveraging non-fungible accelerators vs highly generic CPU, and the unique reliability, efficiency, and cost obstacles those workloads face.  Our solutions should where possible provide automation that benefits other serving workloads like web services or stateful databases, be usable as primitives by multiple ecosystem projects, and compose well into the workflows of those deploying models to production.

How does that read?

Author: thank you! rephrased a little bit. WDYT?

Reviewer 5: Looks good

  and WGs around needs for identified requirements. Do it in partnership
  with existing ecosystem projects like kServe, Seldon, Kaito, and
  others to identify, extract, or implement common shared problems (like Kueue
  abstracted deferred scheduling for multiple batch frameworks).
- Specific areas of improvement include:
  - Directly improve key kubernetes workload controllers when used with
    accelerators and the most common inference serving frameworks and model
    servers.
  - Explore new projects that improve orchestration, scaling, and load balancing
    of inference workloads and compose well with other workloads on Kubernetes
  - Being able to run serving workloads safely while giving up
    available slack capacity to batch frameworks

### Out of scope

- Training and batch inference, which are covered by WG Batch.
- Ability to describe the workflows for serving workloads is out of scope,
  Kubernetes will offer building blocks to MLOps platforms to build those.

## Stakeholders

Stakeholders in this working group span multiple SIGs that own parts of the
code in core kubernetes components and addons.

> - SIG Apps as an primary SIG

Reviewer 7: nit
```suggestion
- SIG Apps as a primary SIG
```

- SIG Architecture
- SIG Node
- SIG Scheduling
- SIG Autoscaling
- SIG Network
- SIG Instrumentation
- SIG Storage

## Deliverables

The list of deliverables include the following high level features:

- To SIG Apps:
  - Ability to express the model serving workloads with easy to understand logical
    objects with the ability to scale to multi-host 
- To SIG Scheduling and Autoscaling
  - Faster scaling up and down
  - Ability to preempt workloads
- To SIG Node:
  - Runtime support for Pods preemption
  - Runtime support for devices partitioning

## Roles and Organization Management

> This wg adheres to the Roles and Organization Management outlined in [wg-governance]

Reviewer 7: nit
```suggestion
This WG adheres to the Roles and Organization Management outlined in [wg-governance]
```

and opts-in to updates and modifications to [wg-governance].

[wg-governance]: /committee-steering/governance/wg-governance.md

> Additionally, the wg commits to:

Reviewer 7: nit
```suggestion
Additionally, the WG commits to:
```

- maintain a solid communication line between the Kubernetes groups and the wider CNCF community;
- submit a proposal to the KubeCon/CloudNativeCon maintainers track;

## Timelines and Disbanding

> As a first mandate, the wg will define a roadmap in the first quarter of operation.

Reviewer 7: nit:
```suggestion
As a first mandate, the WG will define a roadmap in the first quarter of operation.
```

# WG Serving Charter

This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
the Roles and Organization Management outlined in [wg-governance].

[Kubernetes Charter README]: /committee-steering/governance/README.md

## Scope

> Discuss and enhance serving workloads on Kubernetes, specifically focused on
> hardware-accelerated AI/ML inference.
> 
> The working group will focus on the novel challenges of compute-intensive online
> inference. Scenarios solving use cases involving non-fungible accelerators will
> be prioritized over solution against generic CPU. However all improvements
> should, where possible, benefits other serving workloads like web services or
> stateful databases, be usable as primitives by multiple ecosystem projects, and
> compose well into the workflows of those deploying models to production.
> 
> The Working Group Batch has similar scope. The difference in scope by simplified
> definition, is that the Serving WG  will generally concentrate on the workloads
> where Pods are running with `restartPolicy=Always`, while WG Batch will
> generally be looking at Pods with the `restartpolicy=OnFailure`. There are edge
> cases to this definition, but it creates an easy enough framework to
> differentiate the scope these two Working Groups.

Reviewer: ```suggestion
Discuss and enhance serving workloads on Kubernetes, specifically focusing on hardware-accelerated AI/ML inference.

The working group will focus on the novel challenges of compute-intensive online inference. Scenarios solving use cases involving non-fungible accelerators will be prioritized over solutions against generic CPU. However, all improvements should, where possible, benefit other serving workloads like web services or stateful databases, be usable as primitives by multiple ecosystem projects, and compose well into the workflows of those deploying models to production.

The Working Group Batch has a similar scope. The difference in scope by a simplified definition is that the Serving WG will generally concentrate on the workloads where Pods are running with restartPolicy=Always, while WG Batch will generally be looking at Pods with the restartPolicy=OnFailure. There are edge cases to this definition, but it creates an easy enough framework to differentiate the scope of these two Working Groups.
```

Some suggestions to the text

Author: Applied

> differentiate the scope these two Working Groups.

Reviewer 3: I really like this improvement. The section on the job types makes it quite clear. 👍

Author: thanks!
